### PR TITLE
Device Template creation.

### PR DESCRIPTION
**Pull request by tehAgitto**

### CODE REVIEW

**Review by tzietkowski — DISMISSED** [review]

File: vmngclient/tests/test_templates.py [hunk 1]
```diff
@@ -61,238 +63,175 @@ def setUp(self):
         self.task = TaskStatus("Success", "success", [])
 
     @patch("vmngclient.session.vManageSession")
-    def test_templates_success(self, mock_session):
+    @patch("vmngclient.response.vManageResponse")
```
Comment (cicharka) [inline vmngclient/tests/test_templates.py:66]: Order of the @patch'es should be like in decorators - in this case I believe that `mock_session` argument in test is mocking `vmngclient.response.vManageResponse`

File: vmngclient/api/templates/device_template/device_template.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            trim_blocks=True,
+            lstrip_blocks=True,
+            undefined=DebugUndefined,
+        )
+        template = env.get_template(self.payload_path.name)
+        output = template.render(self.dict())
+
+        ast = env.parse(output)
+        if meta.find_undeclared_variables(ast):
+            print(meta.find_undeclared_variables(ast))
```
Comment (cicharka) [inline vmngclient/api/templates/device_template/device_template.py:59]: please use logger.

File: vmngclient/api/template_api.py [hunk 3]
```diff
@@ -56,50 +54,129 @@ def __init__(self, name):
...
 
-    def get(self, name: str) -> Template:
-        """
+        templates = self.session.get(url=endpoint, params=params)
+        return templates.dataseq(DeviceTemplateInfo)
 
-        Args:
-            name (str): Name of template.
+    @overload
+    def get(self, template: type) -> DataSequence[TemplateInfo]:
```
Comment (cicharka) [inline vmngclient/api/template_api.py:91]: unknow type hint

File: vmngclient/api/template_api.py [hunk 4]
```diff
@@ -56,50 +54,129 @@ def __init__(self, name):
...
+
+    def attach(self, template, device, name=None, **kwargs):
+        if isinstance(template, CLITemplate):
+            return self._attach_cli(template, name, device)
+
+        if isinstance(template, DeviceTemplate):
+            return self.attach_feature(template.name, device, **kwargs)
+
+        if template is DeviceTemplate and name:
+            self.attach_feature(name, device, **kwargs)
```
Comment (cicharka) [inline vmngclient/api/template_api.py:123]: probably missing if

File: vmngclient/api/template_api.py [hunk 5]
```diff
@@ -56,50 +54,129 @@ def __init__(self, name):
...
+
+        raise NotImplementedError()
+
+    # TODO list of devices
+    def attach_feature(self, name: str, device: Device, **kwargs):
```
Comment (cicharka) [inline vmngclient/api/template_api.py:128]: as `**kwargs` can be important for the user, I would ask to add docstring explaining why they should be dict that specifies`"device_specific_vars"`. Not needed if this is mentioned in documentation on wiki

File: vmngclient/api/template_api.py [hunk 6]
```diff
@@ -205,58 +283,103 @@ def create(
...
+
+    @overload
+    def create(self, template: CLITemplate) -> str:
+        ...
+
+    def create(self, template):
+        if isinstance(template, list):
+            return [self.create(t) for t in template]
+
+        template_id: Optional[str] = None
```
Comment (cicharka) [inline vmngclient/api/template_api.py:360]: Optional not needed?

**Comment by renukavinay** [comment]
+1

**Review by renukavinay — COMMENTED** [review]
secondary-address is out of ip

File: vmngclient/api/templates/payloads/cisco_vpn_interface_ethernet/feature/cisco_vpn_interface_ethernet.json.j2 [hunk 7]
```diff
@@ -34,29 +45,18 @@
...
-      "vipType": "constant",
-      "vipValue": "{{ interface_name.type.value }}{% if interface_name.number is not none %}{{ interface_name.number }}{% endif %}",
-      "vipVariableName": "vpn_if_name"
-    },
-    "description": {
-      "vipObjectType": "object",
-      "vipType": "ignore",
-      "vipVariableName": "vpn_if_description"
-    },
-    "secondary-address": {
```
Comment (renukavinay) [inline vmngclient/api/templates/payloads/cisco_vpn_interface_ethernet/feature/cisco_vpn_interface_ethernet.json.j2:51]: secondary-address is not in order

**Review by renukavinay — COMMENTED** [review]
Please look at the order in the payload for interfaces.

File: vmngclient/api/template_api.py [hunk 8]
```diff
@@ -56,48 +54,141 @@ def __init__(self, name):
...
     def __init__(self, session: vManageSession) -> None:
         self.session = session
 
-    @property
-    def templates(self) -> List[Template]:
-        """
+    def _get_feature_templates(
+        self, summary: bool = True, offset: Optional[int] = None, limit: Optional[int] = None
+    ) -> DataSequence[FeatureTemplateInfo]:
+        """In a multitenant vManage system, this API is only available in the Provider view."""
```
Comment (tomaszwarda) [inline vmngclient/api/template_api.py:72]: Do we want to block it by checking `session_type`? Or at least check if session type is provider and if not log warning.

File: vmngclient/api/template_api.py [hunk 9]
```diff
@@ -245,58 +366,102 @@ def create(
...
+    def create(self, template):
+        if isinstance(template, list):
+            return [self.create(t) for t in template]
+
+        template_id: Optional[str] = None  # type: ignore
+        template_type = None
+
+        # exists = self.get(type(template)).filter(name=template.name)
+        # if exists:
+        #     raise AlreadyExistsError(f"Template [{template.name}] already exists.")
```
Comment (tomaszwarda) [inline vmngclient/api/template_api.py:447]: Intentionally left?

File: vmngclient/api/template_api.py (hunk 8 above)
Comment (tehAgitto) [reply vmngclient/api/template_api.py:72]: Probably we need to discuss that internally. There are several endpoints working only on provider view and we do nothing about that.

File: vmngclient/api/templates/payloads/cisco_vpn_interface_ethernet/feature/cisco_vpn_interface_ethernet.json.j2 (hunk 7 above)
Comment (tzietkowski) [reply vmngclient/api/templates/payloads/cisco_vpn_interface_ethernet/feature/cisco_vpn_interface_ethernet.json.j2:51]: secondary-address is in ip
image

**Review by tomaszwarda — APPROVED** [review]

**Review by sbasan — APPROVED** [review]
approving conditionally, adding task to update unittests #185

**Pull request merged**